util/election/runner.go

```diff
@@ -0,0 +1,237 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/util"
+)
+
+// EventType is a type of event occurred to an election instance.
+type EventType byte
+
+// EventType possible values.
+const (
+	BecomeMaster     EventType = iota // The instance believes to be the master.
+	NotMaster                         // Lost mastership along the way.
+	NotMasterResign                   // Voluntarily resigned from mastership.
+	NotMasterTimeout                  // Resigned due to status update timeout.
+)
+
+// Config is a set of parameters for a continuous master election process.
+type Config struct {
+	// PreElectionPause is the maximum interval to wait before starting election.
+	PreElectionPause time.Duration
+	// RestartInterval is the time the instance will wait before attempting to
+	// Start election after an error returned by the previous Start call.
+	RestartInterval time.Duration
+	// MasterCheckInterval is the interval between checks that the instance still
+	// holds mastership.
+	MasterCheckInterval time.Duration
+	// MasterTTL is the time after which the instance will resign if encountered
+	// no successful IsMaster calls.
+	MasterTTL time.Duration
+	// MasterHoldInterval is the minimum interval to hold mastership for during
+	// flawless operation.
+	MasterHoldInterval time.Duration
+	// ResignOdds gives the chance of resigning mastership after each check
+	// interval, as the N for 1-in-N.
+	ResignOdds int
+}
+
+// ShouldResign returns whether the instance can voluntarily give up on
+// mastership according to the config.
+func (c *Config) ShouldResign(elected, now time.Time) bool {
+	passed := now.Sub(elected)
+	if passed < c.MasterHoldInterval {
+		// Always hold onto mastership for a minimum interval to prevent churn.
+		return false
+	}
+	// Roll the bones.
+	return c.ResignOdds <= 0 || rand.Intn(c.ResignOdds) == 0
+}
+
+// Event describes whether the instance has become or stopped being the master.
+// Each Event *must* be Ack-ed when acted upon by the subscriber.
+type Event struct {
+	// Type denotes the type of event occurred to the election instance.
+	Type EventType
+	// done notifies the Runner back when the Event has been acted upon.
+	done chan struct{}
+}
+
+// Ack notifies the Runner that it can follow up with a new election round.
+// This allows the downstream client act on the fact of resigning before the
+// instance re-enters election, e.g. they may want to gracefully/forcefully
+// finish the outstanding work.
+func (e *Event) Ack() {
+	if e.done != nil {
+		close(e.done)
+	}
+}
+
+// Runner coordinates master election for a single instance. Reports mastership
```

Comment:
... coordinates election participation for a ...

Reply:
Done.